qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -69,40 +67,46 @@ def run(self, dag):
-                params = run[0].op.params
-                # Remove single identity gates
-                if len(params) > 0 and np.array_equal(run[0].op.to_matrix(), identity_matrix):
-                    dag.remove_op_node(run[0])
-                    continue
-                if isinstance(run[0].op, U3Gate):
-                    param = float(params[0])
-                    if math.isclose(param, 0, rel_tol=0, abs_tol=1e-12) or math.isclose(
-                        param, np.pi / 2, abs_tol=1e-12, rel_tol=0
-                    ):
-                        single_u3 = True
-                    else:
-                        continue
-                else:
-                    continue
-
             new_circs = []
             operator = run[0].op.to_matrix()
             for gate in run[1:]:
                 operator = gate.op.to_matrix().dot(operator)
-            for decomposer in self.basis:
+            for decomposer in self._decomposers:
                 new_circs.append(decomposer._decompose(operator))
             if new_circs:
                 new_circ = min(new_circs, key=len)
-                if len(run) > len(new_circ) or (single_u3 and new_circ.data[0][0].name != "u3"):
+
+                # is this run all in the target set and also uncalibrated?
+                rewriteable_and_in_basis_p = all(
+                    g.name in self._target_basis and not dag.has_calibration_for(g)
+                    for g in run
+                )
+                # does this run have uncalibrated gates?
+                uncalibrated_p = any(not dag.has_calibration_for(g) for g in run)
+                # does this run have gates not in the image of ._decomposers _and_ uncalibrated?
+                uncalibrated_and_not_basis_p = any(
+                    g.name not in self._target_basis and not dag.has_calibration_for(g)
+                    for g in run
+                )
+
+                if (rewriteable_and_in_basis_p and len(run) < len(new_circ)):
+                    # NOTE: This is short-circuited on calibrated gates, which we're timid about
+                    #       reducing.
+                    warnings.warn(f"Resynthesized {run} and got {new_circ}, "
+                                  f"but the original was native and the new value is longer.  This "
+                                  f"indicates an efficiency bug in synthesis.  Please report it by "
+                                  f"opening an issue here: "
+                                  f"https://github.com/Qiskit/qiskit-terra/issues/new/choose",
+                                  stacklevel=2)
+                # if we're outside of the basis set, we're obligated to logically decompose.
+                # if we're outside of the set of gates for which we have physical definitions,
+                #    then we _try_ to decompose, using the results if we see improvement.
+                if uncalibrated_and_not_basis_p or (uncalibrated_p and len(run) >= len(new_circ)):
```

Comment:
i don't really understand what the "_p" means in these names but the overall logic looks good to me!

Reply:
"This returns / is a boolean", with p short for predicate. I did it out of mindless old habit, though now that you mention it I recognize that it's idiosyncratic. I don't mind dropping the `_p` if you'd prefer.